apps/hannk/interpreter/transforms.cpp

```diff
@@ -63,79 +65,89 @@ void remove_dead_ops(OpGroup *root) {
     }
 }
 
+}  // namespace
+
+void remove_dead_ops(OpGroup *root) {
+    std::set<Tensor *> root_outputs;
```

Comment:
I think we should just iterate through the outputs whenever we need to check if a tensor is in the set of outputs. It'd be simpler and the performance upside is certainly negligible. It might even be a lot faster for such small sets.

Maybe just add a helper function `is_output(const Op *, const Tensor *)` and call it whenever querying this set. I'm pretty sure I already had that helper function somewhere at least at one time.